Download models in the background

ahcorde opened the pull request:
Signed-off-by: ahcorde <ahcorde@gmail.com>
# 🎉 New feature

Related to https://github.com/ignitionrobotics/ign-gazebo/issues/1260

it requires https://github.com/gazebosim/sdformat/tree/ahcorde/sdf9/worldName

## Summary

This PR allows to download models in the background, the simulation will not start until all models are downloaded and inserted in the world.

![downloadParallel](https://user-images.githubusercontent.com/1933907/165631060-77529ebd-b76d-44cf-a2e6-63249bacea2b.gif)

## Checklist
- [ ] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

chapulina reviewed (commented):
I took an initial look, but I don't fully follow all of the code yet. Would you mind adding some comments in the code, or in the PR description, explaining the idea of the implementation?

From what I've understood so far, it looks like we may be doing some duplicate work loading the SDF file, is that correct? Another question I have is whether we're correctly handling nested includes and models spawned at runtime that have includes inside them.

ahcorde commented:
The idea of the implementation is the following:

 1. Read the sdf, this will call the method `FetchResource` from `ServerPrivate` class. In this class I launch a thread for each `<include>` tag. This might raise some error because we are not able to find some models. I filtered the error in the `Server`class constructor.
 2. While the models are being downloaded the simulation runner is launched, this will allow to control the camera in the 3DScene, the simulation can't be started until all models are downloaded.
 3. When all models are downloaded the thread which download the last model will take care of reload the sdf file and create the models, actor or lights that are new in the file. this time we will be able to load the data. This is happening in `ServerPrivate::FetchResource`. I needed to use the LevelManager to insert the models in the scene through the SimulationRunners, that's why I added `Addmodel`, `AddLight` and `AddActor` to `SimulationRunner` and `LevelManager` classes.


**NOTE**: There is a special case when the visual or collision mesh uri contains an URL, in this case the model is loaded fine we don't have any report that a specific model is not loaded.

ahcorde commented:
@chapulina I changed the implementation a little bit.

Right now:
 - I started the simulation with an empty world
 - Download the models in the background without changing `ServerPrivate::FetchResource`
 - When models are downloaded, then I added them to the runner

jennuine commented on src/Server.cc:
```diff
@@ -148,31 +152,127 @@ Server::Server(const ServerConfig &_config)
+    std::string filePath = systemPaths.FindFile(this->dataPtr->config.SdfFile());
+
+    std::string worldName;
+    auto errors2 = this->dataPtr->sdfRoot.GetWorldName(filePath, worldName);
```
This is giving me compilation errors. I see https://github.com/gazebosim/sdformat/pull/1027 targeting Garden, it doesn't look like there are breaking changes. Should it target `sdf9` so that it works with this PR? Or are you planning on closing this PR in favor of https://github.com/gazebosim/gz-sim/pull/1484?

ahcorde replied:
you need this branch https://github.com/gazebosim/sdformat/tree/ahcorde/sdf9/worldName

Both PR are doing the same thing, I want to include this on citadel, but I have some issues here that are not happing in Garden.

jennuine commented:
Testing with `ign gazebo fuel.sdf`, I am able to interact with the GUI while the models are loading but they do not load one by one like in your gif. They all show up in the scene all at once when I see the message: `[Msg] Download models in parallel has finished. Now you can start the simulation`

nkoenig commented on src/Server_TEST.cc:
```diff
@@ -82,7 +82,7 @@ TEST_P(ServerFixture, DefaultServerConfig)
 
   EXPECT_EQ(3u, *server.EntityCount());
-  EXPECT_TRUE(server.HasEntity("default"));
+  EXPECT_TRUE(server.HasEntity("empty_world"));
```
Is this correct? I wouldn't expect the word name to change.

nkoenig commented on src/SimulationRunner.cc:
```diff
@@ -718,13 +733,45 @@ bool SimulationRunner::Run(const uint64_t _iterations)
+        isInitialRunOfSimulationSet = true;
+        this->SetPaused(!this->serverConfig.RunOption());
+
+        // this->initialEntityCompMgr.CopyFrom(this->entityCompMgr);
```
Remove these comments?

nkoenig commented on src/SimulationRunner.cc:
```diff
@@ -42,6 +42,21 @@ using namespace gazebo;
 
 
+//////////////////////////////////////////////////
+/// \brief Structure to temporarily store plugin information for reset
```
Did some reset work accidentally get merged into this PR?

nkoenig reviewed (commented):
Just an initial set of comments. I still need to look at the Server and try it out.

ahcorde commented on include/ignition/gazebo/Server.hh:
```diff
@@ -120,6 +120,20 @@ namespace ignition
+      /// ServerConfig and choose it should download models in the background
+      /// \param[in] _config Server configuration parameters. If this
+      /// parameter is omitted, then an empty world is loaded.
+      public: Server(bool _downloadInParallel,
```
The `Init` method is called from the constructor and it's using `downloadInParallelp, we need to set this parameter first

ahcorde commented:
Closing this PR in favour of this other one in Garden https://github.com/gazebosim/gz-sim/pull/1484
